services/src/projects/projectdb.rs

```diff
@@ -1,37 +1,42 @@
-use crate::error::Result;
 use crate::projects::project::{
     CreateProject, Project, ProjectId, ProjectListOptions, ProjectListing, UpdateProject,
 };
 
 use async_trait::async_trait;
 
-use super::{LoadVersion, ProjectVersion};
+use super::{error::ProjectDbError, LoadVersion, ProjectVersion};
 
```

Comment:
Well, you've updated the error handling.

But which scenarios should we try out and see what we get? It is quite hard to go into Python or so and try out errors. But I guess you already tried out different things and could share them.

Reply:
> But which scenarios should we try out and see what we get?

Here is a test where you can see what you get: https://github.com/geo-engine/geoengine/blob/80f2649acc5e1021ec9aa1130515e646d5e6b5d1/services/src/handlers/projects.rs#L1118

I don't think we have to adjust anything in Python or the Web UI because we have no real error handling other than showing it to users as a string or a toast.

Reply:
Cmon… you don't have to adjust something in Python. I don't care about this.

But what are the scenarios that we should look at now to see the new errors and assess them?
I thought you would have some scripts that make these errors happen in order to check the code.

How did you try it out? Otherwise, I have to "write" this from scratch.

Reply:
As I said in the daily scrum: There is a test for deletion which shows the error. Other than that, I have not written any scripts to gather errors messages.

Reply:
I looked at several errors via swagger-ui:

1. 
`GET /project/{project}` with wrong id
> {
  "error": "LoadLatestProjectVersion",
  "message": "Could not load latest project version: Accessing project 36574dc3-560a-4b09-9d22-d5945f2b8093 failed"
}

Not that clear to me as a user. Why latest version? And why did it fail?

2.  `DELETE /project/{project}` 

> {
  "error": "DeleteProject",
  "message": "Could not delete project: Accessing project 36574dc3-560a-4b09-9d22-d5945f2b8093 failed"
}

Also not clear why. I mean, I see that it failed in the first place, so this does not give me additional info why.

3. The same goes for `PATCH /project/{project}` 

4.  `PATCH /project/{project}` when I throw in some garbage JSON

 > {
  "error": "BodyDeserializeError",
  "message": "expected `,` or `}` at line 14 column 3"
}

If I'm not familiar with serde, this is hard to understand

5. `PATCH /project/{project}` when I want to colorizer to be of type `foo`

>{
  "error": "BodyDeserializeError",
  "message": "data did not match any variant of untagged enum VecUpdate at line 19 column 3"
}

Not that helpful either :smile: 

6. `GET /project/{project}/{version}` when I insert a correct project but a wrong version number

> {
  "error": "LoadProjectVersion",
  "message": "Could not load project version: Project 185255cc-7ac8-4cd9-a01f-3b0b1c8fc6df does not exist"
}

I guess this is not the right error?

7.  `GET /projects`  does not terminate, no clue why.

---

Any idea to trigger other errors of you?

Reply:
1. This error occurs when you do not have the permission to access the project. As the error is defined in Free, I cannot mention anything about permissions directly. Thus this is only mentioned in the source. I added the source error to the display now. ed302b34e47790c51f42fe0bb386e00bffd38e6a
2. dito
3. dito
4. I changed the error message to mention that the error is in the user input. As I said, I do not know how to give more helpful error messages. 5ab94207a7fb5052b5eb772f86e56cf8a7ad07cc
5. dito
6. I changed the error handling to output a different error message when trying to access a specific version of a project 221cf73b394369c422dcc0fddfc808d6cfce3c16
7. Works for me. I did

> POST http://localhost:3030/api/project
Authorization: Bearer bd8bbaca-1e78-4f75-8348-f2108b2a51f9
Content-Type: application/json
>
> {
  "name": "Test",
  "description": "Foobar",
  "bounds": {
    "spatialReference": "EPSG:4326",
    "boundingBox": {
      "lowerLeftCoordinate": {
        "x": -180,
        "y": -90
      },
      "upperRightCoordinate": {
        "x": 180,
        "y": 90
      }
    },
    "timeInterval": {
      "start": 0,
      "end": 0
    }
  }
}
>
>###
>
>GET http://localhost:3030/api/projects?permissions=["Owner"]&filter=None&order=NameAsc&offset=0&limit=1
Authorization: Bearer bd8bbaca-1e78-4f75-8348-f2108b2a51f9
Content-Type: application/json

and got

> HTTP/1.1 200 OK
content-length: 152
connection: close
content-type: application/json
date: Wed, 06 Sep 2023 14:56:43 GMT
>
> [
  {
    "id": "2e6e8528-68c8-46fe-adad-84bb0e7e783d",
    "name": "Test",
    "description": "Foobar",
    "layerNames": [],
    "plotNames": [],
    "changed": "2023-09-06T14:51:46.189Z"
  }
]

Reply:
1-3. The message is kind of ok now. Is the error code helpful, though?
4.-5.  Much better, but we should tackle the untagged enum stuff separately.
6. Looks still wrong for me, I add a correct project but a wrong version… "Project 480ee428-2d84-4f60-a01d-32868afd4b03 does not exist"
7. /projects does not work in swagger for me

Reply:
> 1-3. The message is kind of ok now. Is the error code helpful, though?

idk, which one would you like better? We do not use them anyway though so I would keep it simple

> 7. /projects does not work in swagger for me

it was a problem with the ProjectFilter which apparently never worked via swagger. I removed the filter because it wasn't actually applied in the DB and not mapped in Python or UI.

f84f871b3c09ccbbc0cb4e073637f20f1e2b5028

Reply:
The project version thing works :+1: .

Now, when I just perform the request on /projects, I get

```
{
  "error": "UnableToParseQueryString",
  "message": "Unable to parse query string: missing field `order`"
}
```